### PR TITLE
feature/preprocess-train-eval-and-more

### LINKED ISSUE #3: Use CML for CI/CD
https://cml.dev/doc/usage

### LINKED ISSUE #5: Speaker diarisation before annotation
Already in progress in #1, implement a function to apply speaker diarisation to help with annotation. Additional utilities can be developed as well to place the grouped audio samples into a directory with the speaker label as the directory name. Labeling them all at once?

### LINKED ISSUE #6: Write function to check the basics of the data (splits and amount)
We don't care how many events are used so much as there are enough events to have entirely new events in the holdout set. And how long events are doesn't matter as long as the number of sentences per person * their average length is generally enough to train a model.

It may actually be better to have a 20% train set and an 80% holdout set so that we can even more so try to ensure the model performs well on new data.

### PR DESCRIPTION
**Pull request by evamaxfield**
### Link to Relevant Issue

This pull request resolves #5, resolves #6, resolves #3

### Description of Changes

A continuation of #1

Opening this up for comments as I am going to start training a full model for Seattle 2021 tomorrow or Friday. This PR basically implements everything needed for larger prototyping and training. I am sure some refactoring will happen down the road, some utility functions will be added and more but as it is right now, I can use this library to both quickly annotate and train models!

Because of how large the file diff is I might recommend simply viewing the repo from the branch view: https://github.com/CouncilDataProject/speakerbox/tree/feature/dataset-expansion

The workflow image doesn't render in the README but that's because the image isn't on the main branch yet, here is the workflow image: https://github.com/CouncilDataProject/speakerbox/blob/feature/dataset-expansion/docs/_static/images/workflow.png

Note the massive caviate that this library currently only works on Ubuntu due to upstream dependencies not yet building for other platforms.

You can somewhat ignore the confusion matrix on this repo / PR -- the training loss drops to 0 on CPU sometimes and it feels like a bug in the huggingface trainer API but I am not entirely sure. When I train on GPU I never run into problems with the model completely failing.

(For example here is the confusion matrix from running the tests locally with GPU enabled)
validation-confusion
(Yes, that is really with as little data as is in the stored test resources zipfile :tada: )

In this case, I am using the confusion matrix here to simply check that training ran / as a proof of concept for how to report back to instance maintainers the results of training. (I am thinking they will open a PR on their instance repos to add their annotation files and we can write a bot they can message to kick off a job to train and then report back something similar to this PR but with more helpful non-technical comments)

Please leave any and all comments!

### CODE REVIEW

**Comment by evamaxfield** [comment]
I can't mark you as reviewer for some reason @kristopher-smith but if you want to leave comments please do!

**Comment by dphoria** [comment]
Not sure how well I can contribute to this :sweat_smile: but will do my best and take a close look this weekend.

**Comment by isaacna** [comment]
Will try to take a look at this sometime this week/weekend!

**Comment by evamaxfield** [comment]
> Not sure how well I can contribute to this :sweat_smile: but will do my best and take a close look this weekend.

I think looking over the README and saying "this does not make sense" is good enough for me

**Comment by evamaxfield** [comment]
Fully trained model. I'm going to add more Juarez examples

**Review by isaacna — COMMENTED** [review]
I'm not very qualified to comment on ML/data science things, but the process described in the README generally makes sense to me! 

I had a few questions (my bad if it's somewhere in the PR, the diff is just a lot):
-  What's the general accuracy/precision you've seen from runs on larger data sets? The confusion matrix from the local run seems like the labels were generally pretty accurate
- What's the split between test/training/validation data sets that you used? Is this hardcoded somewhere or can this be customized?

**Comment by evamaxfield** [comment]
Good questions!

> * What's the general accuracy/precision you've seen from runs on larger data sets? The confusion matrix from the local run seems like the labels were generally pretty accurate

I have only tried training the full model twice. The first time it fully trained, it succeeded in training but it didn't save to S3 properly. That model reported a 98% accuracy but I don't have the train, test, validation split counts, or any more details other than this line from the log: `Loading best model from trained-speakerbox/checkpoint-4128 (score: 0.9816171573198348).`

Since split designation is random but with a few rules, I will add some more data for Juarez and lewis and expect ~94% - 99% accuracy.

> * What's the split between test/training/validation data sets that you used? Is this hardcoded somewhere or can this be customized?

The above full model was trained, tested, and evaled on:
```
           train_counts  test_counts  valid_counts
gonzalez          3503          713           928
herbold           2969          470           947
juarez             387          216           311
lewis              964          237           266
morales            700          306           447
mosqueda          2818          530          1007
pedersen          1132          176           351
sawant             584          682           564
strauss           1213           51           264
```

[See here for more dataset prep info](https://github.com/CouncilDataProject/speakerbox/blob/feature/dataset-expansion/speakerbox/preprocess.py#L415). Currently you can provide a seed to the splitting function but no explicit conversation ids. I will add an issue to the repo for such a feature but rn I don't think it's needed because that function has been working pretty well.

File: README.md [hunk 1]
```diff
@@ -17,20 +17,171 @@ Speaker Annotation for Transcripts using Audio Classification
...
+    ├── ...
+    └── 53998-62821.wav
+```
 
-seattle_2021_ds_dir = datasets.unpack_seattle_2021_proto(clean=True)
-seattle_2021_ds_summary_stats = datasets.summarize_annotation_statistics(
-    seattle_2021_ds_dir / "annotations"
+We leave it to you as a user to then go through these directories and remove any audio
+clips that were incorrectly clustered together as well as renaming the sub-directories
+(from "SPEAKER_00", "SPEAKER_01", "SPEAKER_02", etc. to "A", "B", "C", etc.).
```
Comment (tohuynh) [inline README.md]: `renaming the sub-directories` -> `renaming the sub-directories to their correct speaker labels` (or something along that line) to make it explicit?

File: README.md [hunk 2]
```diff
@@ -17,20 +17,171 @@ Speaker Annotation for Transcripts using Audio Classification
...
+and for utilizing multiple audio annotation schemes to train such a model.
+
+## Workflow
+
+⚠️ `speakerbox` currently only works on Ubuntu due to upstream dependencies ⚠️
+
+### Diarization
+
+We quickly generate an annotated dataset by first diarizing (or clustering based
+on the features of speaker audio) portions of larger audio files and splitting each the
```
Comment (tohuynh) [inline README.md:62]: Just curious, how accurate is the pre-trained model for this step in clustering speakers? It might give a clue as to how much cleaning up is needed.

File: speakerbox/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+# Data expections and cleaning
+
+
+def prepare_dataset(
+    dataset: pd.DataFrame,
+    test_and_valid_size: float = 0.4,
+    equalize_data_within_splits: bool = False,
+    n_iterations: int = 100,
+    seed: Optional[int] = None,
+) -> Tuple[DatasetDict, pd.DataFrame]:
```
Comment (tohuynh) [inline speakerbox/preprocess.py:421]: I think we want two things
- roughly uniform distribution of labels (in each of the audio files, and thus also in each of the dataset splits - train, test, valid)
- the same distribution of labels across train, test, and valid sets

Seems like `equalize_data_within_splits` is going to throw out some data points?
Maybe we could ask user (in the README) to use audio files where each speaker has roughly the same amount of speaking time or as close as possible?

**Review by tohuynh — COMMENTED** [review]
The README and workflow image makes sense to me.

File: README.md (hunk 1 above)
Comment (tohuynh) [reply README.md]: Also, if an audio was incorrectly clustered, we wouldn't want the user to move it to the correct sub-directory?

Comment (evamaxfield) [reply README.md]: I personally just deleted the audio files that were incorrectly clustered but it would likely be useful as a "negative" example.

File: README.md (hunk 2 above)
Comment (evamaxfield) [reply README.md:62]: From this metric: https://github.com/pyannote/pyannote-audio#benchmark

I think the closest is probably the VoxConverse evaluation set of 12.8% error

But it also depends on how distinct each speaker is and how much cross talking there is.

File: speakerbox/preprocess.py (hunk 3 above)
Comment (evamaxfield) [reply speakerbox/preprocess.py:421]: > * roughly uniform distribution of labels (in each of the audio files, and thus also in each of the dataset splits - train, test, valid)

From my large model tests this doesn't seem to have much of an effect / we may want to purposfully bias the dataset. I.e. if Gonzalez talks a majority of the time because she is council president, the model should be confident that she is talking a majority of the time too. But this is exactly why this parameter was added. I can attempt to train a large model with this parameter turned on.

> Seems like `equalize_data_within_splits` is going to throw out some data points?

Yes it does so at random, per split.

> Maybe we could ask user (in the README) to use audio files where each speaker has roughly the same amount of speaking time or as close as possible?

Yep, I can at least add that documentation, the best example of this for us was using the council briefing meetings which are the "what am I doing this week in council" where everyone goes around and shares. The sitting council president naturally has higher speaking times for any meeting, including briefings, however.

Comment (tohuynh) [reply speakerbox/preprocess.py:421]: > we may want to purposfully bias the dataset. I.e. if Gonzalez talks a majority of the time because she is council president, the model should be confident that she is talking a majority of the time too. But this is exactly why this parameter was added. I can attempt to train a large model with this parameter turned on.

I don't think that's needed. If we don't want the uniform distribution of labels, then I think the same distribution across train, test, and valid is good enough.

**Comment by evamaxfield** [comment]
@tohuynh 

I trained a bunch more models because machine learning is deterministic :upside_down_face: 

best of the bunch:
* equalized labels across splits:
validation-confusion

  * **Accuracy:** 0.9709784411276948
  * **Precision:** 0.9723741343113816
  * **Recall:** 0.9709784411276949
  
[training log](https://github.com/JacksonMaxfield/phd-infrastructures/runs/6462014299?check_suite_focus=true)

* random good performer without equalization
validation-confusion

  * **Accuracy:** 0.9732294962277926
  * **Precision:** 0.9734652240956935
  * **Recall:** 0.9755533085841637

[training log](https://github.com/JacksonMaxfield/phd-infrastructures/runs/6461243451?check_suite_focus=true)

* random good performer without equalization num 2
validation-confusion

  * **Accuracy:** 0.9766917293233083
  * **Precision:** 0.9726282428102516
  * **Recall:** 0.9809027891876685

[training log](https://github.com/JacksonMaxfield/phd-infrastructures/runs/6461088449?check_suite_focus=true)

---

I feel like we could use any of these :shrug:

**Comment by tohuynh** [comment]
> I feel like we could use any of these 🤷

Sick! 

Curious to see the predicted probabilities of each data point for the valid set or just the cross-entropy loss of each model for the valid set.

**Comment by evamaxfield** [comment]
> > I feel like we could use any of these 🤷
> 
> Sick! 
> 
> Curious to see the predicted probabilities of each data point for the valid set or just the cross-entropy loss of each model for the valid set.

Those confusion matrices are from predictions on the holdout validation set 😀

The validation loss should be reported in the logs? I think... If not I think we can add an issue to the repo to add later?

**Comment by tohuynh** [comment]
> The validation loss should be reported in the logs? I think... If not I think we can add an issue to the repo to add later?

Not sure if it is. Since the accuracy, precision, and recall scores were so similar across these models, I was curious if the loss was similar too.

**Comment by evamaxfield** [comment]
> Not sure if it is. Since the accuracy, precision, and recall scores were so similar across these models, I was curious if the loss was similar too.

Ahhhh yea we don't have loss for the validation but if you open those logs and use the search logs box for "eval_loss" you can jump to different epoch loss reports

**Comment by tohuynh** [comment]
> if you open those logs and use the search logs box for "eval_loss" you can jump to different epoch loss reports

That's for the test set right? The best I saw was ~0.4.

I think we want at least 0.2 on the validation set?

Cross-Entropy = 0.00: Perfect probabilities.
Cross-Entropy < 0.02: Great probabilities.
Cross-Entropy < 0.05: On the right track.
Cross-Entropy < 0.20: Fine.
Cross-Entropy > 0.30: Not great.
Cross-Entropy > 1.00: Terrible.
Cross-Entropy > 2.00 Something is broken.
(https://machinelearningmastery.com/cross-entropy-for-machine-learning/)

Since the model is going to have to predict `Unknown` if the highest predicted probability for a real-world data point is too low (low confidence on this data point). I think it would be good to choose a model with low cross-entropy loss on the validation set -- it means the predicted probability distribution is very similar to the true probability distribution of a data point).

**Comment by evamaxfield** [comment]
I want to change the accuracy, precision, and recall to weighted average instead of "macro" average as I think it is right now. I will also add in computing the loss as well :+1: 

Will rerun a few models to see what happens.

**Comment by evamaxfield** [comment]
Will also try to generate and store the training curves as well

**Comment by tohuynh** [comment]
Something of [note](https://stats.stackexchange.com/questions/258166/good-accuracy-despite-high-loss-value):

> Your loss might be hijacked by a few outliers (very wrong predictions), check the distribution of your loss function on individual samples of your validation set. If there are a cluster of values around the mean then you are overfitting. If there are just a few values very high above a low majority group then your loss is being affected by outliers :)

Hopefully, it's just some bad outliers that made `eval_loss` so high.

**Comment by evamaxfield** [comment]
Well now I am confused...

Here is the most recent model including the validation loss:
validation-confusion

* **Accuracy:** 0.9763440860215054
* **Precision:** 0.9772270405532067
* **Recall:** 0.9763440860215054
* **Validation Loss:** 15.127711990008887

Logs: https://github.com/JacksonMaxfield/phd-infrastructures/runs/6478246169?check_suite_focus=true

**Comment by evamaxfield** [comment]
I am wondering if I should just run more epochs?

**Comment by tohuynh** [comment]
I don't think that would improve anything, training loss is already approaching zero -- pretty much can't minimize training loss anymore. After that, training would be just picking up idiosyncrasies/noises of the data, instead of the general pattern of the data.

**Comment by tohuynh** [comment]
Could you examine the predicted probabilities of each datapoint in the validation set?

**Comment by tohuynh** [comment]
I think we need to see plots for training loss and eval_lost as training happens.

Overall, not sure about what conclusions to draw regarding what training learned from the data, if the accuracy, precision, and recall scores are good, but cross-entropy loss is terrible.

**Comment by evamaxfield** [comment]
From a friend:

> you could have some bonkers outliers

:joy: seems accurate

**Comment by evamaxfield** [comment]
After a bit more discussion and thinking I have one idea as to what is happening and one decision.

1. I agree with my friend Greg that I think the reason our cross entropy is high is because some predictions for Gonzalez when the true speaker was Herbold are just really confident in the prediction. I.e. It predicts the speaker as Gonzalez with 95% confidence all 14 of those mis-classifications. I have a weird hypothesis as to why that may be happening however which is related to not the voice behind the audio, but the words. When I was annotating this data, I noticed and remembered that Herbold (and Juarez briefly) acted as interim Council Presidents while Gonzalez was on maternity leave. When I was annotating, there were a couple of meetings I used in the annotation set which included Herbold as the interim Council President and I have a hypothesis that the model partially learned which words are associated to which individuals on the councils as well as which waveforms. The larger more general hypothesis would be something like "speech recognition models can be trained more accurate by punishing / weighting predictions by ngram diversity" or something similar (Imagine if our training set had both Gonzalez and Herbold saying: "Will the clerk please call the roll?"). For now, advisors and I are going to shelf that idea simply due to time. But likely come back to it when we have more models trained for more instances in an attempt to understand if we can predict "roll" on the council by speech alone.

2. If you recall back to how I am applying these models to the transcript, I am chunking the audio of the sentences of the transcript, and for each sentence chunk predicting the speaker and then averaging the speaker for the whole sentence. With that in mind, I am going to semi-roll back some of the last three commits and ship this for now. With 97.65% accuracy I hope everyone is okay with that.

**Comment by tohuynh** [comment]
Besides the confidently wrong classifications you mentioned above, the rest are confidently (or good enough for us) right?

> I have a weird hypothesis as to why that may be happening however which is related to not the voice behind the audio, but the words. 

These confidently wrong classifications for audio segments of Herbold include words also spoken by Gonzalez (in the training set)? If Herbold and Gonzalez both said XYZ in the training set, wouldn't the model be able to pick up the difference?

>With that in mind, I am going to semi-roll back some of the last three commits and ship this for now. With 97.65% accuracy I hope everyone is okay with that.

👍

**Comment by evamaxfield** [comment]
> These confidently wrong classifications for audio segments of Herbold include words also spoken by Gonzalez (in the training set)? If Herbold and Gonzalez both said XYZ in the training set, wouldn't the model be able to pick up the difference?

Right but their may be more examples of Gonzalez using such terminology than Herbold using such terminology and so the model sees the terminology as the defining factor and not the voice.

**Comment by evamaxfield** [comment]
But to be clear. I don't know if that is truly the case or not. Purely speculating. The fact that everything else is incredibly accurate and says to me that its picking up voice. Especially because I know there were some meetings in the dataset where Juarez and Mosqueda were interim president.
